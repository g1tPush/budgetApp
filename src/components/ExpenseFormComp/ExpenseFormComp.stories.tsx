import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import ExpenseFormComp from './ExpenseFormComp'

import { StoreProvider, expenseList } from '../../index'

export default {
    title: 'Expense Form Composition',
    component: ExpenseFormComp
} as ComponentMeta<typeof ExpenseFormComp>

const Template: ComponentStory<typeof ExpenseFormComp> = (args) => <ExpenseFormComp {...args} />

export const CreateComp = Template.bind({})

CreateComp.args = {
    screenType: 'create'
}

CreateComp.decorators = [
    (story) => (
        <StoreProvider value={expenseList}>
            {story()}
        </StoreProvider>
    )
]

export const DeleteComp = Template.bind({})

DeleteComp.args = {
    screenType: 'edit'
}

DeleteComp.decorators = [
    (story) => (
        <StoreProvider value={expenseList}>
            {story()}
        </StoreProvider>
    )
]
