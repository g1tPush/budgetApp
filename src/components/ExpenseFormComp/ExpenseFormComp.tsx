import React from 'react'
import { Container } from '../../shared/global'
import PageHeader from '../PageHeader/PageHeader'
import ExpenseForm from '../ExpenseForm/ExpenseForm'
import { Wrap } from '../../shared/global'

interface ExpenseFormCompProps {
    /**
     * Does this button delete anything?
     */
    screenType: 'edit' | 'create'
}


const ExpenseFormComp: React.FC<ExpenseFormCompProps> = ({ screenType }) => {
    return (
        <>
            <Wrap>
                <Container>
                    <PageHeader headerType={screenType} />
                </Container>
            </Wrap>
            <Container style={{ display: 'flex', flexDirection: 'column', gap: '25px' }}>
                <ExpenseForm formType={screenType} />
            </Container>
        </>
    )
}

export default ExpenseFormComp