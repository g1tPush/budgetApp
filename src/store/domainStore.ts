import { makeObservable, observable, makeAutoObservable } from 'mobx'
import moment from 'moment'
import { IExpense, SortType, DateType } from '../types/types'

// {
//     id: '123ew7t23ieuwd',
//     content: 'Iphone',
//     amount: 1000,
//     date: moment(),
//     note: 'This is a good choise'
// }

export class ExpensesStore {
    expenses: Expense[] = []
    searchText: string = ''
    type: SortType = 'Date'
    startDate: DateType = null
    endDate: DateType = null
    filterredArr: Expense[] = []

    constructor() {
        makeObservable(this, {
            expenses: observable,
            searchText: observable,
            type: observable,
            startDate: observable,
            endDate: observable,
            filterredArr: observable
        })
    }

    get expensesList() {
        return this.expenses.slice().filter((item) => {
            if (item.date) {
                const date = moment(item.date).format('MM-DD-YYYY')
                if (this.startDate === null && this.endDate === null) {
                    return item
                } else if (this.startDate !== null && this.endDate !== null) {
                    return new Date(date) <= new Date(this.endDate.format('MM-DD-YYYY')) && new Date(date) >= new Date(this.startDate.format('MM-DD-YYYY'))
                } else if (this.startDate === null) {
                    return new Date(date) <= new Date(this.endDate?.format('MM-DD-YYYY') || '')
                } else if (this.endDate === null) {
                    return new Date(date) >= new Date(this.startDate?.format('MM-DD-YYYY') || '')
                }
            }
            return 1
        }).filter((item) => {
            return item.content.toLowerCase().includes(this.searchText.toLowerCase())
        }).sort((a, b) => {
            if (b.date && a.date && this.type === 'Date') {
                return b.date?.valueOf() - a.date?.valueOf()
            } else if (this.type === 'Amount') {
                return +(b.amount) - +(a.amount)
            }
            return 1
        })
    }

    updateDate(startDate: DateType, endDate: DateType) {
        this.startDate = startDate
        this.endDate = endDate
    }

    updateType(type: 'Date' | 'Amount') {
        this.type = type
    }

    updateSearchText(text: string) {
        this.searchText = text
    }

    createExpense(expense: IExpense) {
        this.expenses.push(new Expense(expense))
        return expense
    }

    updateExpense(expense: IExpense) {
        let index = this.expenses.findIndex(el => {
            return el.id === expense.id
        })

        this.expenses[index] = expense
    }

    removeExpense(expense: Expense | undefined) {
        if (expense) {
            const index = this.expenses.findIndex((el) => {
                return el.id === expense.id
            })
            this.expenses.splice(index, 1)
        }
    }

    getExpense(expenseId: string | undefined) {
        if (expenseId === undefined) {
            return
        }
        return this.expenses.slice().find(el => {
            return el.id === expenseId
        })
    }
}

export class Expense {
    id: string | null = null
    content: string = ''
    date: DateType = null
    note: string;
    amount: number = 0

    constructor(store: IExpense) {
        makeAutoObservable(this, {
            id: false,
            content: false,
            date: false,
            amount: false
        })


        this.id = store.id
        this.amount = store.amount
        this.content = store.content
        this.note = store.note
        this.date = store.date
    }
}