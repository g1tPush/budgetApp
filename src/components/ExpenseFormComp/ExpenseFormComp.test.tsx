import React from 'react'
import renderer from 'react-test-renderer'
import ExpenseFormComp from './ExpenseFormComp'
import { BrowserRouter } from 'react-router-dom'
import { StoreProvider } from '../../index'
import { ExpensesStore } from '../../store/domainStore'

jest
    .useFakeTimers()
    .setSystemTime(new Date('01-01-2022'))

it('renders edit form', () => {
    const tree = renderer
        .create(
            <StoreProvider value={new ExpensesStore()}>
                <BrowserRouter>
                    <ExpenseFormComp screenType='edit' />
                </BrowserRouter>
            </StoreProvider>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})

it('renders create form', () => {
    const tree = renderer
        .create(
            <StoreProvider value={new ExpensesStore()}>
                <BrowserRouter>
                    <ExpenseFormComp screenType='create' />
                </BrowserRouter>
            </StoreProvider>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})