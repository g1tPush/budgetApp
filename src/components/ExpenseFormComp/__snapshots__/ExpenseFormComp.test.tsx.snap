// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders create form 1`] = `
Array [
  <div
    className="sc-bczRLJ kigNbx"
  >
    <div
      className="sc-gsnTZi dkDfzv"
    >
      <p
        className="sc-hKMtZM cpxqda"
      >
        Create Expense
      </p>
    </div>
  </div>,
  <div
    className="sc-gsnTZi dkDfzv"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "gap": "25px",
      }
    }
  >
    <form
      className="sc-ksZaOG"
      data-testid="test-form"
      onSubmit={[Function]}
    >
      <div
        className="sc-hAZoDl hCUrIr"
      >
        <input
          className="sc-eCYdqJ cslQzD"
          data-testid="description"
          onChange={[Function]}
          placeholder="Description"
          required={true}
          type="text"
          value=""
        />
        <input
          className="sc-eCYdqJ cslQzD"
          data-testid="amount"
          onChange={[Function]}
          pattern="[0-9.]+"
          placeholder="Amount"
          required={true}
          type="text"
          value=""
        />
        <div
          className="SingleDatePicker SingleDatePicker_1"
        >
          <div>
            <div
              className="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
            >
              <div
                className="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-date"
                  aria-label="Date"
                  autoComplete="off"
                  className="DateInput_input DateInput_input_1"
                  disabled={false}
                  id="date"
                  name="date"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Date"
                  readOnly={false}
                  required={false}
                  type="text"
                  value="01-01-2022"
                />
                <p
                  className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-date"
                >
                  Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
            </div>
          </div>
        </div>
        <textarea
          className="sc-fnykZs iZABqP"
          data-testid="note"
          name="note"
          onChange={[Function]}
          placeholder="Add note"
          value=""
        />
      </div>
      <button
        className="sc-dkzDqf eUEFT"
        mode={0}
        type="submit"
      >
        Save Expense
      </button>
    </form>
  </div>,
]
`;

exports[`renders edit form 1`] = `
Array [
  <div
    className="sc-bczRLJ kigNbx"
  >
    <div
      className="sc-gsnTZi dkDfzv"
    >
      <p
        className="sc-hKMtZM cpxqda"
      >
        Edit Expense
      </p>
    </div>
  </div>,
  <div
    className="sc-gsnTZi dkDfzv"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "gap": "25px",
      }
    }
  >
    <form
      className="sc-ksZaOG"
      data-testid="test-form"
      onSubmit={[Function]}
    >
      <div
        className="sc-hAZoDl hCUrIr"
      >
        <input
          className="sc-eCYdqJ cslQzD"
          data-testid="description"
          onChange={[Function]}
          placeholder="Description"
          required={true}
          type="text"
          value=""
        />
        <input
          className="sc-eCYdqJ cslQzD"
          data-testid="amount"
          onChange={[Function]}
          pattern="[0-9.]+"
          placeholder="Amount"
          required={true}
          type="text"
          value=""
        />
        <div
          className="SingleDatePicker SingleDatePicker_1"
        >
          <div>
            <div
              className="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
            >
              <div
                className="DateInput DateInput_1"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-date"
                  aria-label="Date"
                  autoComplete="off"
                  className="DateInput_input DateInput_input_1"
                  disabled={false}
                  id="date"
                  name="date"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Date"
                  readOnly={false}
                  required={false}
                  type="text"
                  value="01-01-2022"
                />
                <p
                  className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                  id="DateInput__screen-reader-message-date"
                >
                  Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
              </div>
            </div>
          </div>
        </div>
        <textarea
          className="sc-fnykZs iZABqP"
          data-testid="note"
          name="note"
          onChange={[Function]}
          placeholder="Add note"
          value=""
        />
      </div>
      <button
        className="sc-dkzDqf eUEFT"
        mode={0}
        type="submit"
      >
        Save Expense
      </button>
      <button
        className="sc-dkzDqf kpebWh"
        data-testid="removeExpense"
        mode={1}
        onClick={[Function]}
        style={
          Object {
            "marginTop": "15px",
          }
        }
      >
        Remove Expense
      </button>
    </form>
  </div>,
]
`;
